### ISSUE TITLE
NewRelic doesn't work when buildpack is used as Cloud Foundry task

### ISSUE THREAD

**Issue by mjanser** [cloudfoundry/php-buildpack#207]
#### What version of Cloud Foundry and CF CLI are you using? (i.e. What is the output of running `cf curl /v2/info && cf version`?
```json
{
   "name": "",
   "build": "",
   "support": "https://support.pivotal.io",
   "version": 0,
   "description": "",
   "authorization_endpoint": "https://...",
   "token_endpoint": "https://...",
   "min_cli_version": "6.7.0",
   "min_recommended_cli_version": "6.11.2",
   "api_version": "2.65.0",
   "app_ssh_endpoint": "...:2222",
   "app_ssh_host_key_fingerprint": "...",
   "app_ssh_oauth_client": "ssh-proxy",
   "routing_endpoint": "https://...",
   "logging_endpoint": "wss://...",
   "doppler_logging_endpoint": "wss://..."
}
```
`cf version 6.26.0+9c9a261.2017-04-06`

#### What version of the buildpack you are using?
Buildpack version 4.3.31-1493150241

#### If you were attempting to accomplish a task, what was it you were attempting to do?
I'm running the php-buildpack as a Cloud Foundry task with NewRelic enabled. Example PHP script:
```bash
php -r 'var_dump(newrelic_start_transaction("test"));'
```

#### What did you expect to happen?
NewRelic data is collected and sent to NewRelic.

#### What was the actual behavior?
No data is collected, the above script outputs `false` because it can't connect to the NewRelic daemon.
Because no php-fpm is started in this case and [some timing issues](https://discuss.newrelic.com/t/php-agents-tries-to-connect-before-daemon-is-ready/48160) the NewRelic daemon is not ready early enough.

Can the NewRelic daemon be [started on boot](https://docs.newrelic.com/docs/agents/php-agent/advanced-installation/starting-php-daemon-advanced) for non-web runs?

Another issue is that when the container is destroyed after the task has finished the NewRelic daemon isn't terminated correctly, so it can't send its data.

**Comment by sclevine** [comment]
There isn't a huge difference between launching an app and running a task -- both use the lifecycle launcher.

@dmikusa-pivotal, do we do something in the default start command that makes new relic work?

**Comment by dmikusa-pivotal** [comment]
>No data is collected, the above script outputs false because it can't connect to the NewRelic daemon.
>Because no php-fpm is started in this case and some timing issues the NewRelic daemon is not ready early enough.

I don't really see how PHP-FPM matters here.  Is it that when running with PHP-FPM it's a long running process, so there's sufficient time to get the daemon up and working?

Right now we enable NewRelic through php.ini settings.  We don't start the agent manually.  I'm not a NR expert, but my understanding is that by adding the settings to php.ini, NR will automatically start itself in the background.  It appears that this process is too slow under certain circumstances and the app goes away before NR can report back its data.

I'm wondering if it would be sufficient to add a pause before & after the script to give it enough time to work properly?

>Can the NewRelic daemon be started on boot for non-web runs?

I briefly looked at the docs that were linked above and it appears NR allows you to start the daemon manually.  Presumably, the extension in PHP will just connect to the running daemon instead of starting up a new one.

Seems like it would be feasible to do this with the build pack.  Definitely going to require some reworking of the NR extension for the build pack.  If we do go down this road, I don't think we should care if it's a web or stand-alone app.  We should just use this method of starting NR across the board.  That should make implementation simpler.

I'm still a little skeptical that this will help.  If the app is so short lived that the NR daemon doesn't have time to start, the same thing might happen if we try to start the daemon manually.  I don't think the build pack has any enforceable order for the processes that it kicks off (i.e. it's not PHP then Nginx, it fires them both up in parallel).  If we add the NR daemon and it's slow, it could potentially still start after the app has started and completed.  Again, we could probably work around this, but it would be a little hacky and take more work.  

>Another issue is that when the container is destroyed after the task has finished the NewRelic daemon isn't terminated correctly, so it can't send its data.

There's not much we can do to prevent the container from being cleaned up.  If the container is properly shutdown (i.e. `cf stop` or something like that), then the app should get a SIGTERM and have 10 seconds to shutdown.  If the app does not nicely shutdown within 10s, it will be SIGKILL'd.  This behavior cannot be changed.

If the application crashes unexpectedly (either PHP-FPM or Nginx / HTTPD), the container is going to go away immediately.  There might be some small chance for us to inject a delay here because of the way the PHP build pack uses a process manager (if PHP-FPM or Nginx/HTTPD die, the build pack's process manager kills the other process so that the container gets restarted).  It would have to be a very small window though because the app is going to be down at this point and waiting will prevent CF from restarting the app.  We need to exit out completely before CF can restart the container.

It would be good to know more about the impact of this missing data.  How many seconds of data are being lost?  Is this critical to debugging a crash?  If so, how?

**Comment by mjanser** [comment]
> I don't really see how PHP-FPM matters here. Is it that when running with PHP-FPM it's a long running process, so there's sufficient time to get the daemon up and working?

In agent mode the NewRelic daemon is started when PHP-FPM starts up as you said correctly. This way the daemon is most probably ready when the first request can be handled. Even if it's not ready for the first request, that would result in only one missed request.

On other hand when no PHP-FPM is running (which is the case for CF tasks), the daemon is started when the PHP command is executed. As mentioned in the [NewRelic forum](https://discuss.newrelic.com/t/php-agents-tries-to-connect-before-daemon-is-ready/48160) there is a timing issue with that. The result is that the agent can't connect to the daemon when calling `newrelic_start_transaction()` and even if the other calls would work, the daemon has no started transaction. Therefore no data is sent to NewRelic for entire CF task.

> I'm wondering if it would be sufficient to add a pause before & after the script to give it enough time to work properly?

A pause before running the PHP command won't help here because the daemon will only be started when PHP runs.

> I'm still a little skeptical that this will help. If the app is so short lived that the NR daemon doesn't have time to start, the same thing might happen if we try to start the daemon manually. I don't think the build pack has any enforceable order for the processes that it kicks off (i.e. it's not PHP then Nginx, it fires them both up in parallel). If we add the NR daemon and it's slow, it could potentially still start after the app has started and completed. Again, we could probably work around this, but it would be a little hacky and take more work.

This is the case for web apps, right? For those I don't see a real problem with that because if one of the services (PHP-FPM or nginx) is not ready, the app can't handle any request anyway.

But for CF tasks I would assume that there is a difference between container start up scripts and the command I'd like to run, in terms that the command is run **after** the container started up completely. This means starting the daemon manually on container start up should work (the official init script for the NewRelic daemon waits for one second to give the daemon time to start up).

Is my assumption wrong or did I miss something?

>  If the container is properly shutdown (i.e. cf stop or something like that), then the app should get a SIGTERM and have 10 seconds to shutdown. If the app does not nicely shutdown within 10s, it will be SIGKILL'd. This behavior cannot be changed.

Do you know if this is also the case for CF task containers?

And which processes get the SIGTERM signal? Does this include the NewRelic daemon?

> It would be good to know more about the impact of this missing data. How many seconds of data are being lost? Is this critical to debugging a crash? If so, how?

We currently don't have any NewRelic data for all CF tasks, which includes regular cron jobs and queue workers. This is quite critical for us.

**Comment by dmikusa-pivotal** [comment]
>On other hand when no PHP-FPM is running (which is the case for CF tasks), the daemon is started when the PHP command is executed. As mentioned in the NewRelic forum there is a timing issue with that. The result is that the agent can't connect to the daemon when calling newrelic_start_transaction() and even if the other calls would work, the daemon has no started transaction. Therefore no data is sent to NewRelic for entire CF task.

Bear with me here as I'm by no means an expert with NewRelic.  So what you're saying is that when your task runs, it calls `newrelic_start_transaction()` to start capturing metrics.  It's at this point that the NR extension starts up the agent.  Because it takes time to start the agent and the NR extension is not waiting for this to happen (sounds like an NR bug to me), no metrics are captured or reported for the entire task.

Would it be possible to fake out the extension?  Ex:  call `newrelic_start_transaction()`, sleep a couple seconds, stop the transaction (I'm assuming there's something like `newrelic_stop_transaction()`).  Then call `newrelic_start_transaction()` again.  At this point, hopefully, the agent would be ready.  Similarly, when your script is done call `newrelic_stop_transation()`, sleep for 60s so metrics are sent out (or perhaps look up the agent's PID and send it a SIGTERM), then exit.

>This is the case for web apps, right? For those I don't see a real problem with that because if one of the services (PHP-FPM or nginx) is not ready, the app can't handle any request anyway.
>But for CF tasks I would assume that there is a difference between container start up scripts and the command I'd like to run, in terms that the command is run after the container started up completely. This means starting the daemon manually on container start up should work (the official init script for the NewRelic daemon waits for one second to give the daemon time to start up).

Yes, disregard what I said here.  I wasn't thinking about things in the frame of a task (i.e. `cf tasks`).  In that case, the build pack isn't going to start anything.  The only thing that runs is the task command you specify.  In fact, I'm not sure that the build pack can start anything for you here.  If you need something started ahead of time, I think your command would have to include that.

Normally when an app runs, the build pack get's to set the command to run your app (through the `release` script).  For a task, I don't believe that this happens (@sclevine - Can you confirm this?).  My understanding is that CF takes your droplet, starts a container with it and runs the command you gave.  It's like using `cf push -c`, where it completely overrides the output of the `release` script.

>Do you know if this is also the case for CF task containers?

I haven't tested this, but I'm assuming that it will behave the same.

>And which processes get the SIGTERM signal? Does this include the NewRelic daemon?

It's always the main process that gets the SIGTERM.  In the case of a PHP web app, we have a `start` script that runs and kicks of PHP-FPM & Nginx / HTTPD.  This is what receives the SIGTERM and is responsible for properly stopping PHP-FPM & Nginx / HTTPD.

In the case of a task, I think it would be the command that you run.  So if you run `php -r 'var_dump(newrelic_start_transaction("test"));'` then the PHP process is going to get the SIGTERM.  If you were to start NewRelic in the background, I'm not really sure if it would end up getting the signal.  That's something that I'd have to test to figure out.

**Comment by mjanser** [comment]
No problem, I'm also no NewRelic expert, just trying to find out why this doesn't work and how to fix it :)

> Would it be possible to fake out the extension? Ex: call newrelic_start_transaction(), sleep a couple seconds, stop the transaction (I'm assuming there's something like newrelic_stop_transaction()). Then call newrelic_start_transaction() again. At this point, hopefully, the agent would be ready. Similarly, when your script is done call newrelic_stop_transation(), sleep for 60s so metrics are sent out (or perhaps look up the agent's PID and send it a SIGTERM), then exit.

That could work out I think, but is definitely not my favorite solution. I agree with you that it looks like a bug in NewRelic, but I didn't get a clear response from them yet.

> In the case of a task, I think it would be the command that you run. So if you run php -r 'var_dump(newrelic_start_transaction("test"));' then the PHP process is going to get the SIGTERM. If you were to start NewRelic in the background, I'm not really sure if it would end up getting the signal. That's something that I'd have to test to figure out.

I understand this when the task is being terminated from outside. But what happens if the task finishes (successful or failed)? Does then any termination happen, too?

**Comment by mjanser** [comment]
> In fact, I'm not sure that the build pack can start anything for you here. If you need something started ahead of time, I think your command would have to include that.

Couldn't `_service_command()` in [lib/php/extension.py](https://github.com/cloudfoundry/php-buildpack/blob/2b2fa0414be3ff0b4c593e54b7b130c340e00338/lib/php/extension.py#L81) be used for starting such services?

```python
def _service_commands(self):
    if is_web_app(self._ctx):
        return {
            'php-fpm': (
                '$HOME/php/sbin/php-fpm',
                '-p "$HOME/php/etc"',
                '-y "$HOME/php/etc/php-fpm.conf"',
                '-c "$HOME/php/etc"')
        }
    else:
        app = find_stand_alone_app_to_run(self._ctx)
        return {
            'php-app': (
                '$HOME/php/bin/php',
                '-c "$HOME/php/etc"',
                app)
}
```

**Comment by dmikusa-pivotal** [comment]
>I understand this when the task is being terminated from outside. But what happens if the task finishes (successful or failed)? Does then any termination happen, too?

My understanding is that when your task exits, the platform won't send a SIGTERM.  The platform only cares about the command you tell it to run (i.e. your task command).  Anything else that would happen to be running in the container is not monitored by the platform.  Thus when the command you tell it to run exits, it's going to assume there is nothing else running and reap the container.

>Couldn't _service_command() in lib/php/extension.py be used for starting such services?

No.  Service commands are gathered up by the build pack and used to build the config file for the process manager.  The start command generated by the build pack starts the process manager which in turn starts the service commands.

In a task, your command is going to run not the command generated by the build pack.  The only way you could use this is if the command you executed was the service manager and the service manager then kicked off your command.

Something hacky that you could do would be to start the NR agent in a `.profile.d/` or `.profile` script that you include with the app.  That would run before your command in the task.  I don't think it is going to behave properly though when your command exits (I suspect it's just going to get a SIGKILL).  You could test it and see though.

**Comment by mjanser** [comment]
I now have a really hacky workaround for this. Not happy with it, but it seems to work.

On command start it's waiting for the NewRelic daemon to be ready, something like this:
```php
$count = 0;
while (!newrelic_set_appname(ini_get('newrelic.appname')) && $count < 5) {
    ++$count;
    sleep(1);
}
newrelic_start_transaction(ini_get('newrelic.appname'));
```

And at the end of the command it terminates the daemon, something like this:
```php
newrelic_end_transaction();
sleep(5);
posix_kill((int) file_get_contents(ini_get('newrelic.daemon.pidfile')), SIGTERM);
sleep(5);
```

**Comment by sclevine** [comment]
@dmikusa-pivotal To answer you question above, a task is very similar to launch: `.profile` and `.profile.d` scripts are sourced, and the env should be the same. The exception is that the start command is not executed.

**Comment by sclevine** [comment]
Closing due to inactivity.
